src/libraries/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.netcoreapp.cs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Runtime.Serialization;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    internal abstract partial class BaseCodePageEncoding : EncodingNLS, ISerializable
+    {
+        internal static unsafe EncodingInfo [] GetEncodings(CodePagesEncodingProvider provider)
+        {
+            lock (s_streamLock)
+            {
+                s_codePagesEncodingDataStream.Seek(CODEPAGE_DATA_FILE_HEADER_SIZE, SeekOrigin.Begin);
+
+                int codePagesCount;
+                fixed (byte* pBytes = &s_codePagesDataHeader[0])
+                {
+                    CodePageDataFileHeader* pDataHeader = (CodePageDataFileHeader*)pBytes;
+                    codePagesCount = pDataHeader->CodePageCount;
+                }
+
+                EncodingInfo [] encodingInfoList = new EncodingInfo[codePagesCount];
+
+                Span<byte> pCodePageIndexBytes = stackalloc byte[sizeof(CodePageIndex)]; // 40 bytes
+                CodePageIndex* pCodePageIndex = (CodePageIndex*) Unsafe.AsPointer(ref pCodePageIndexBytes.GetPinnableReference());
```

Comment:
You can avoid the `Unsafe.AsPointer` and the `stackalloc` by writing instead:

```cs
CodePageIndex codePageIndex = default;
```

This also helps ensure proper data alignment, which the runtime doesn't necessarily guarantee on its own when it sees _stackalloc_. Then, if you need to turn the entire struct into a `Span<byte>`:

```cs
DoSomethingWith(new Span<byte>(&codePageIndex, Unsafe.SizeOf<CodePageIndex>());
```

And if you need to turn the name into a LPCWSTR:

```cs
DoSomethingWith(&codePageIndex.CodePageName);
```

This is also a good opportunity to clean up `s_codePagesDataHeader` so that it's typed as `CodePageDataFileHeader` instead of `byte[]`, which allows removing some of the unsafe code used to access it. But that can be split into a separate PR if we so choose.

Reply:
I did that. Thanks for the suggestion.